[FIX] account: reconcile in multicompany

mreficent opened the pull request:
**How to reproduce the issue/feature this PR addresses:**

 - Enable multicompany option for demo user.
 - Create two companies, A and B, and make company B child of company A.
 - Make the two companies allowed for demo user.
 - Play with the demo user in `account` module, changing companies, and may you find something unexpected.

 But this PR is just about the reconcile case. As an example, in bank statement reconcile:
 - The demo user, using company A, creates an invoice for a partner.
 - The demo user, using company B, creates an invoice for same partner.
 - Using company A, demo user creates a bank statement and try to reconcile.
- What the widget will show to the demo user are the two invoices (they have same name/reference), and that's wrong.

![selection_007](https://user-images.githubusercontent.com/25005517/36214204-2323f61e-11a9-11e8-998d-e12f2b903c55.png)

This example is fixed with the proposed addition of the `company_id` in the search domain, in the `get_move_lines_for_reconciliation` method of `account.bank.statement.line`.

The changes proposed in `get_statement_line_for_reconciliation_widget`, `prepare_move_lines_for_reconciliation_widget` and `_renderCreate` are to restrict company wise the fields in the account reconcile model form of the bank statement reconcile widget.

The other two changes are just for cases where you need to consider also the company_childs.

**Current behavior before PR:** The reconciliations are not working correctly in a multicompany case as the example shown above. And it's difficult to do custom modules to implement this changes, because it's very ugly to override this large methods with calling super and also to override the javascript file somehow.

**Desired behavior after PR is merged:** Reconciliations are working as expected. Changes in `.xml` files are not suggested because version is stable.


--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

mreficent commented:
Well, the runbot fails in the javascript test but it's still operative. My knowledge on javascript is limited, I don't know if the javascript file fix could be done in a more "correct" way.

@qdp-odoo what do you think about this?

blaggacao commented:
@mreficent Do I understan your claim correctly?

"A parent company should not be able to do reconcile own payments with the invoices of a child company!"

I would contest that this is rather usual, but only if they belong to the same accounting company (share the same chart of accounts).

Can you clarify?

jbeficent commented:
@blaggacao As I understand it in odoo companies do not share charts of accounts. Each has it's own. So it would not be acceptable for the parent company to pay for a child company's invoice. 

@qdp-odoo @odony this PR is part of our bigger effort to close all the holes required for a user to be able to operate in various companies. See https://github.com/Eficent/multicompany-fixes. In this specific case it was hard for us to simply create an addon module, and instead we are proposing this PR.

@richard-willdooit do you agree with the PR? As you are facing the same issue.

jbeficent commented:
As discussed in https://github.com/OCA/manufacture/pull/254, we might better use the picking_id field in mrp.bom to locate where is the product being manufactured.

blaggacao commented:
@jbeficent Of course, but then this should be labeled rather as a treatment than a fix. Because the underlying problem is that Odoo does not allow for such operations. That level of isolation is just not the case in reality.
At least let's reference that possibility in future or modules... https://github.com/odoo/odoo/pull/19839

qdp-odoo commented:
wait... what? we never wants to consider the child companies: it always have to be restricted to the company of the statement line, isn't it??

jbeficent commented:
@qdp-odoo exactly. The statement should only propose to reconcile with moves that belong to the  company of the statement, and now it shows child companies as well.

qdp-odoo commented:
but the PR seems to continue allowing that
```diff
+        company_ids = self.env.user.company_id
+        company_ids |= self.env.user.company_id.child_ids
+        company_ids = tuple(company_ids.ids)
  
         sql_query = """SELECT stl.id
                         FROM account_bank_statement_line stl
                         WHERE account_id IS NULL AND stl.amount != 0.0 AND not exists (select 1 from account_move_line aml where aml.statement_line_id = stl.id)
-                            AND company_id = %s
+                            AND company_id in %s
                 """
-        params = (self.env.user.company_id.id,)
+        params = (company_ids,)
```
??

mreficent commented:
@qdp-odoo maybe you are right and this case about the child companies is wrong fixed. The intention behind this was to assure the reconciliation was not with the child companies but to show them in case they are considered. As you say, maybe that is wrong done.

But as I pointed at the initial post, this PR has several changes. Just take the ones you feel safe and discard the others. And remember the important fix would be the widget one that maybe @Gorash would find the right implementation 😊

qdp-odoo commented:
ok then, we'll consider this while merging the refactoring of Gorash.

mreficent commented:
@qdp-odoo updated the PR as you requested.

We considered what you said about the child companies case and our position is that we divided the case in the following two subcases:
 - bank reconciliation (**reconciliation_widget_preprocess** method): We changed the `env.user.company` by the statement's company because the user maybe is reconciling journals of child companies. To be sure that all statements have the same company, we added an UserError in the `account_journal_dashboard.py`

 - manual reconciliation (**get_data_for_manual_reconciliation** method): We maintain the addition of the child companies because in a manual reconciliation we consider it makes sense to access to the possible reconciliations of the child companies. As an example, the following images show the manual reconciliation for a company and child company:
![selection_013](https://user-images.githubusercontent.com/25005517/36793763-87412c98-1c9e-11e8-9f6f-8b583b1180b2.png)
![selection_014](https://user-images.githubusercontent.com/25005517/36793767-8a8fa4ec-1c9e-11e8-9496-aecd1bcbd757.png)

We hope the changes proposed are considered. Thanks.

blaggacao commented:
@qdp-odoo In the ongoing paradigm (which obviously is the relevant here), this should not be possible. In a use case where you use companies for rights isolation but share the same CoA (in custom module), this can be a very valid use case. (Thinking about v15 or so... :smile: )

blaggacao commented:
@mreficent 
> bank reconciliation (reconciliation_widget_preprocess method): We changed the env.user.company by the statement's company because the user maybe is reconciling journals of child companies. To be sure that all statements have the same company, we added an UserError in the account_journal_dashboard.py

It also makes sens in a (future/custom) subsidiary model. Holding company receives payments of subsidiary in their bank account or child1 receives payments for child2 which only holding can reconcile.

I personally would appreciate, if those changes could be applied in a way that follows the current paradigm, but does not automatically break (eg UserError) the enhanced scenario...

This would be very muddy to override maintain, then...

qdp-odoo commented:
closed by https://github.com/odoo/odoo/commit/8767ecc693d83564faf084805aeae6ec4793b773

thanks for the contribution :+1:

mreficent commented:
A lot of thanks! @qdp-odoo 

Could be possible to backport to v11?

mreficent commented:
Why in the <code>.js</code> widged the same domain is not included in the `analytic_account_id` field? 🤔

qdp-odoo commented:
I'd prefer not, as it was made after the refactoring of Gorash and could be difficult to do/maintain. 

Moreover, the current situation in stable is already okayish since even if you can select lines from other companies, the system won't allow you to complete the reconciliation if it involves several companies. :disappointed_relieved:

qdp-odoo commented:
> Why in the .js widged the same domain is not included in the analytic_account_id field? :thinking: 

since version saas11.2, analytic accounts/tags can be shared among companies, so I simply let the ir.rules do their duty
